### DISCUSSION TITLE
A Wiki page/Website page/Discussion topic for Early access builds/beta testing?

### DISCUSSION THREAD

**Discussion by pc223** [Flow-Launcher/Flow.Launcher#529] [Ideas]
Should we do something similar like this?  🤔  (like a Wiki page, or Flow's website page or Discussion topic?)

https://github.com/goldendict/goldendict/wiki/Early-Access-Builds-for-Windows

We got automatic CI artifacts so it seems not too hard to do.

I think we only need a couple of main builds:
- A `dev branch` build
- A `next release build` (like `1.8.0`)
- Some `PR build` (builds of a big PRs, e.g. PR #470)

**Why:**
- For people want to test new improvements
- For testing new ideas
- Early feedback, early bug report

**Comment by jjw24** [comment]
i dont know if there are enough interests atm to test builds. the builds can always be accessed via the build results in each PR, perhaps when there are more requests we can put it in a discussion section or wire up the ci for it

**Comment by pc223** [comment]
> i dont know if there are enough interests atm to test builds. the builds can always be accessed via the build results in each PR, perhaps when there are more requests we can put it in a discussion section or wire up the ci for it

I think people will try it out if we provide them the builds clearly (like what that build is for, and where to post testing-results), right now the builds are buried in PR which is not many people aware of and not sure what to do with it.

Early feedback is the whole point of CI/CD stuff, I think we need early feedback on big PRs like PR-#470 or big releases like 1.8.0, or feedback about style (PR-#495) (those tend to be fairly subjective)

And the last release of Flow is like 4 months now. Many users not able to try out all the improvement that exists for months seems a bit ...sad 😅 

I think, for now, we can just keep it really simple, 1 topic (discussion) for it, listed 3 builds: a weekly-dev-builds, a 1.8.0 build, a big PR build. And we ask people to try them out. 

In terms of maintenance, for now we can do manual, weekly edit that topic, later we can write some CI to automate if people like it.

**Comment by jjw24** [comment]
The release pr will constantly have changes merged in, which will be quite manual to keep up with the latest builds in discussion, (I'm afk) I wonder if there is something in AppVeyor to provide a constant link to the latest build of the pr.

Version from builds are still the production release version, for currently it is 1.7.2, I don't know if we need to mark them as pre-release and how.

**Reply by pc223** [reply]
Seems this will link get **latest** good artifacts of the `dev` branch (not check with official docs yet)

Web link:
https://ci.appveyor.com/project/JohnTheGr8/flow-launcher/branch/dev/artifacts

Download link:
https://ci.appveyor.com/api/projects/JohnTheGr8/flow-launcher/artifacts/Output%2FPackages%2FFlow-Launcher-v1.7.2.exe?branch=dev&pr=false

For Pull Requests, I can't find it yet, worst case, we make a branch on the main repo, and grant people commit rights to that branch, not perfect, but automatic => no maintenance cost.


Sources:
https://github.com/appveyor/ci/issues/403#issuecomment-180026470

https://help.appveyor.com/discussions/questions/4501-url-to-the-latest-build-in-the-master-branch

https://www.appveyor.com/docs/packaging-artifacts/#permalink-to-the-last-successful-build-artifact

**Comment by jjw24** [comment]
Also how is this discussion abt adding early access builds in Discussions not already in Discussions 😛

**Comment by pc223** [comment]
> Also how is this discussion abt adding early access builds in Discussions not already in Discussions 😛

Haha hold up, lemme convert this real quick, kind of itchy to try it 😝

**Comment by jjw24** [comment]
manual one will be like this i suppose
https://github.com/Flow-Launcher/Flow.Launcher/discussions/535

**Reply by pc223** [reply]
Yeah, it's cool that we make users aware of how to get appveyor builds. But that download link in the announcement will seem always to download the newest (success, non-pr) `dev` build, in the future, when we do 1.9.0, that link will point to 1.9.0, not 1.8.0, so can be awkward a bit, in the future, for now it's fine. (I can't find the optimal solution yet, make a new branch for 1.8.0 can work, although it can complicate the process)

I think we still need a discussion that listed all the builds that we'd like users  to test like this:
https://github.com/goldendict/goldendict/wiki/Early-Access-Builds-for-Windows

For pre-release builds, PR builds, experiments,... kind of a Beta/Insider program :smile:, we'll ask users to test all those builds for us to get early feedback.

**Reply by jjw24** [reply]
yeah i am aware of the link is pointing to the latest build in dev branch,, would be nice if we can point to pr but just havent had the time to look into it. we can change it to a fixed one once 1.8.0 is out

**Reply by jjw24** [reply]
it is quite manual, so i admit i was a little lazy about it haha

**Reply by pc223** [reply]
What about we make the build list a markdown `.md` file? For normal PRs you just can let PR-dev worry about it, so the list of builds will live in a `.md` file, and people will send changes requests (PR) to that `.md` file.

So we maintain the build list like ...codes, PR -> review->merge like usual :smile:

**Reply by pc223** [reply]
For example, I'd like users to test my weird progress-bar, then I'll make a PR to the `build_list.md` file, and add the appveyor download link in.

**Reply by jjw24** [reply]
must be a better way to do this. your suggestion is still manual if i understand correctly. wouldnt it be even easier just create a discussion like the pre-release one and add the link you want people to test? an announcement called Early Access Builds. You dont need to go through the pr review process as well this way

**Reply by pc223** [reply]
> must be a better way to do this. your suggestion is still manual if i understand correctly. wouldnt it be even easier just create a discussion like the pre-release one and add the link you want people to test? an announcement called Early Access Builds

Announcement for releases is great, but for early access builds/PR builds, I'm thinking of something like only 1 dicussion, or 1 `build_list.md` file or 1 Wiki page, that contains **all** of the _beta_ builds, so users can try all of them at once, instead of have to scour the Discussion to find the builds. 

I think all the builds should be in one place 😄

**Reply by jjw24** [reply]
yeah that's what i mean, just one announcement, call it early access builds and can contain those build links

**Reply by pc223** [reply]
> yeah that's what i mean, just one announcement, call it early access builds and can contain those build links

We have 12 pull requests now, assume we want users to try out those 12 builds,  so that announcement will contains 12 builds?

Those 12 build links will have to update when have new commits tho. Ideally we can have a script auto-update those build links, I think appveyor api support this use case, but don't have time to research on it yet.

**Reply by pc223** [reply]
That's why I mentioned the PRmodel, `build_list.md`, because that oneway to manage all the build links.

**Reply by jjw24** [reply]
well we dont need to let users to test all 12, just the big ones. more importantly however how do we pull the links out and append it to the md?

**Reply by pc223** [reply]
> well we dont need to let users to test all 12, just the big ones. more importantly however how do we pull the links out and append it to the md?

Yeah, I'm hoping that Appveyor support some kind of api to pull links out for PR, maybe have to write a script to do it instead of a simple one-liner URL, I'll research on this later.

**Reply by pc223** [reply]
Seems we have to use these API to automate this:


1. Get history, find the PR build, get the buid version (1.7.2.2000)
https://ci.appveyor.com/api/projects/JohnTheGr8/flow-launcher/history?recordsNumber=3&branch=dev&startBuildId=39801085

2. Get the JobId (m9thia3t5p1kw8ab)
https://ci.appveyor.com/api/projects/JohnTheGr8/flow-launcher/build/1.7.2.2000

3. All artifacts
https://ci.appveyor.com/api/buildjobs/m9thia3t5p1kw8ab/artifacts

4. .exe file direct download link
https://ci.appveyor.com/api/buildjobs/m9thia3t5p1kw8ab/artifacts/Output/Packages/Flow-Launcher-v1.7.2.exe

Ref:
https://kevinoid.github.io/appveyor-swagger/bootprint/
